### ISSUE TITLE
"group=average" is incorrectly produces sum instead of average in API data query

### ISSUE THREAD

**Issue by aldem** [netdata/netdata#6015]
##### Bug report summary

When "group=average" parameter is set in `/api/v1/data` request values are computed incorrectly - sum is reported instead of average.

##### OS / Environment

Debian 9.8

##### Netdata version (output of `netdata -V`)

netdata v1.14.0-43-g8e9268d

Installed using `bash <(curl -Ss https://my-netdata.io/kickstart-static64.sh)`

I use default configuration, the only difference is history which is set to 90000s.

##### Component Name

HTTP API

##### Steps To Reproduce

Retrieve some chart using API `/api/v1/data` with group=average set (as in examples below).

Raw data (without grouping, for 10 arbitrary seconds):
```
$ curl 'http://localhost:19999/api/v1/data?chart=system.cpu&after=1557878600&before=1557878610&dimension=user'
{
 "labels": ["time", "user"],
    "data":
 [
      [ 1557878610, 0.1257862],
      [ 1557878609, 0],
      [ 1557878608, 0.2515723],
      [ 1557878607, 0.250941],
      [ 1557878606, 0],
      [ 1557878605, 0.3778338],
      [ 1557878604, 0.2512563],
      [ 1557878603, 0],
      [ 1557878602, 0.2506266],
      [ 1557878601, 0.1251564]
  ]
}
```

Group by average:
```
$ curl 'http://localhost:19999/api/v1/data?chart=system.cpu&after=1557878600&before=1557878610&dimension=user&group=average&gtime=10'
{
 "labels": ["time", "user"],
    "data":
 [
      [ 1557878610, 1.6331726]
  ]
}
```
This already looks weird (as there are no values above 0), and group by sum produces exactly the same result:
```
$ curl 'http://localhost:19999/api/v1/data?chart=system.cpu&after=1557878600&before=1557878610&dimension=user&group=sum&gtime=10'
{
 "labels": ["time", "user"],
    "data":
 [
      [ 1557878610, 1.6331726]
  ]
}
```

##### Expected behavior

When "group=average" is set, values must be average, but they actually are sum.

**Comment by cakrit** [comment]
https://github.com/netdata/netdata/blob/6ca6d840dd19d5d7e9bacf93e011803ea5861496/web/api/queries/average/average.c#L50
 internal.resampling_group=10 (based on gtime), but internal.resampling_divisor = 1 , instead of the number of points we have. So the sum is divided by 1, instead of 10.

The culprit is here: https://github.com/netdata/netdata/blob/6ca6d840dd19d5d7e9bacf93e011803ea5861496/web/api/queries/query.c#L719
This division doesn't make sense. We essentially do 10*1 / 10 = 1. 

Will discuss with @ktsaou for the fix, because I don't get everything that's being done in there.

**Comment by cakrit** [comment]
Ok, so I was wrong. When you provide the `gtime` directive, you are asking netdata to do resampling, which means to return results in different units. Such a request makes no sense for metrics like system.cpu, but consider what would happen if you were counting http requests. If you got let's say about 2-10 requests/sec with `update every = 1` and wanted to see the average per minute, then you'd provide `gtime=60`. The response you'd get then for that one minute, would be the sum of the requests for each of the 60 seconds. 

What you need to get the average you're looking for is `points=XXX`. Then you'll see that you actually get a proper average.

**Comment by aldem** [comment]
@cakrit Maybe I am mistaken but I see no reason why it makes no sense for system.cpu (or any other % value) - then stddev or median also would make no sense, but those two are computed correctly.

API docs says about gtime: `The grouping number of seconds. This is used in conjunction with group=average to change the units of metrics (ie when the data is per-second, setting gtime=60 will turn them to per-minute)`

So, when there is one minute of 1% CPU usage for every sampled second (which is reflected in raw samples), its average over one minute is still 1% - not 60%, regardless of units and period. Yes, we did use 60% of CPU time within that minute, but on average is was 1% during that minute.

If we count number of events per sample (requests in your example) - yes, then changing of units makes sense and only then average calculation as it is done now is correct (though in this case a simple sum of events over period just equals average per period, that's all) - but any % value is not a number of events/period, so it has to be calculated another way.

In any case, as long as there are stddev/median, those should be consistent with average.

**Comment by cakrit** [comment]
Don't get me wrong, I also find this confusing and I'm happy to continue the discussion. What I understand from the statement `when the data is per-second, setting gtime=60 will turn them to per-minute`, is that it assumes that we talking about metrics with units X/sec. %cpu time/sec isn't a valid unit for the `gtime` directive, but requests/sec and MiB/sec are. 

Regarding the consistency issue, I think you're right. Since the `gtime` param only applies to `group=average` to change the units, perhaps we shouldn't be accepting it for any other request. We could make a modification to return an error if it's provided in the request when `group!=average`, so that it's not as misleading. I am worried though about backwards compatibility here. We can't know how many people use `group=median&gtime=X`.

**Comment by aldem** [comment]
Well, if this will be properly documented and "wrong" usage will return an error - this is good enough, though using "points" to get average over period seems "unnatural" while "gtime" exists.

What is more confusing though is that "average" with "gtime" is actually a "sum of values over period" - perhaps, it should not be called "average" then?

Another option is special handling of absolute values (non-counters) - though this may be complicated and not backward-compatible.

And one more question - average calculation in health checks is using the same code? If so, then health checks might produce wrong results for any non-counters.

**Comment by cakrit** [comment]
> And one more question - average calculation in health checks is using the same code? If so, then health checks might produce wrong results for any non-counters.

health check queries are always done with the equivalent of `gtime=0`, so they don't do that. 

We'll discuss the points you raised internally, they are valid.

**Comment by cakrit** [comment]
I had forgotten about this and we need to improve this area. The operations that one can do on a metric to apply a function across a given time period should depend on a metric's "ALGORITHM". The following is very relevant:

This is what netdata supports:
```
#define RRD_ALGORITHM_ABSOLUTE_NAME                "absolute"
#define RRD_ALGORITHM_INCREMENTAL_NAME             "incremental"
#define RRD_ALGORITHM_PCENT_OVER_DIFF_TOTAL_NAME   "percentage-of-incremental-row"
#define RRD_ALGORITHM_PCENT_OVER_ROW_TOTAL_NAME    "percentage-of-absolute-row"
```

- “absolute” is the same as a prometheus “gauge”, a single value that can’t be summed over time (with another value collected previously). E.g. active connections 
- “incremental” is used for monotonically increasing sums (prometheus “counter”). netdata converts these metrics to /sec units (e.g. requests/sec). Again the result can’t be summed.
- We already use % in several charts and we can create such charts by default for prometheus histograms and summaries, to automatically create always two charts for each histogram or summary: One with raw values and one with percentages.

Opentelemetry tries to add some meaning to raw metrics, to tell the applications using them if they should sum or average a metric across time, and help them understand if they can calculate a rate based on a specific metric. “Monitoring and alerting systems commonly use the data provided through metric events, after applying various aggregations and converting into various exposition formats. “  See https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#overview

The interesting findings from opentelemetry are related to our /api/v1/data “group” attribute, that is used to say how we summarize individual measurements across a given time period: "The median value is considered useful information for non-additive measurements.” In netdata we use average by default in our grouping, not median. We may want to at least provide a UI option to let users select between average and median group selection in the UI settings.

Additive measurements “indicate that only the sum is considered useful information. These are measurements that you would naturally combine using arithmetic addition, usually real quantities of something (e.g., number of bytes).”. By converting all such measurements into non-additive measurements (rates) in netdata, we have essentially lost the ability to sum. The /api/v1/data “group” attribute can’t meaningfully take the values “sum” or “incremental-sum” to aggregate data over a specific period of time. This makes it a bit more difficult for us to provide accurate charts with e.g. requests/hour or requests/day or total requests within a given time period. But we can still do it, by summing over several req/s and changing the unit’s denominator (e.g. sum 60 req/s within a minute and show req/min). See for example what we do for badges (https://github.com/netdata/netdata/blob/master/web/api/badges/web_buffer_svg.c#L362). The API supports this with the `gtime` option discussed above, but it’s not well documented.

So basically the action here is what was already mentioned above, to revisit all possible queries, take into account the collection ALGORITHM (which maps nicely to metric types & unit categories) and support only valid requests for each ALGORITHM. Return error when something is not applicable. 
These need to be well documented in the API spec and also on a separate document in learn.netdata.cloud

**Comment by amoss** [comment]
I wasn't aware of this old discussion when I opened https://github.com/netdata/netdata/issues/9857.

For some metrics we currently interpret the request to calculate an average as a sum. This is wrong and needs to be fixed. If the user requests an average of the values over time in a time-series then we should either:

* Honor the request and calculate the average. The answer may not be meaningful, but this is up to the user to determine.
* Reject the request with an error message.

These are both valid choices. Returning the sum of the values in the time-series instead of the average as a side-effect of the group-size is a bug (although it is in the intended design rather than the implementation).

**Comment by ilyam8** [comment]
Closing this feature request. We will re-evaluate the request internally.
